File: algorithms/sorting/merge-sort/implementation.js
/**
 * Merge Sort Algorithm
 *
 * Efficient divide-and-conquer sorting with guaranteed O(n log n) performance
 *
 * Time: O(n log n) all cases
 * Space: O(n) auxiliary space
 * Stable: Yes
 */


/**
 * Standard recursive merge sort (functional style)
 *
 * @param {number[]} arr - Array to sort
 * @returns {number[]} - New sorted array
 *
 * @example
 * mergeSort([38, 27, 43, 3, 9, 82, 10])
 * // [3, 9, 10, 27, 38, 43, 82]
 */
function mergeSort(arr) {
    // Base case: arrays of 0 or 1 element are already sorted
    if (arr.length <= 1) {
        return arr;
    }

    // Divide: split array in half
    const mid = Math.floor(arr.length / 2);
    const left = arr.slice(0, mid);
    const right = arr.slice(mid);

    // Conquer: recursively sort both halves
    const sortedLeft = mergeSort(left);
    const sortedRight = mergeSort(right);

    // Combine: merge the sorted halves
    return merge(sortedLeft, sortedRight);
}


/**
 * Merge two sorted arrays into one sorted array
 *
 * @param {number[]} left - First sorted array
 * @param {number[]} right - Second sorted array
 * @returns {number[]} - Merged sorted array
 */
function merge(left, right) {
    const result = [];
    let i = 0;  // Left array pointer
    let j = 0;  // Right array pointer

    // Compare elements from left and right, add smaller to result
    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {  // <= ensures stability
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }

    // Copy any remaining elements from left array
    while (i < left.length) {
        result.push(left[i]);
        i++;
    }

    // Copy any remaining elements from right array
    while (j < right.length) {
        result.push(right[j]);
        j++;
    }

    return result;
}


/**
 * In-place merge sort (modifies original array)
 *
 * @param {number[]} arr - Array to sort
 * @param {number} left - Left boundary (default 0)
 * @param {number} right - Right boundary (default arr.length - 1)
 * @returns {number[]} - The same array, sorted
 */
function mergeSortInPlace(arr, left = 0, right = arr.length - 1) {
    if (left >= right) {
        return arr;
    }

    const mid = Math.floor((left + right) / 2);

    // Sort left and right halves
    mergeSortInPlace(arr, left, mid);
    mergeSortInPlace(arr, mid + 1, right);

    // Merge the sorted halves
    mergeInPlace(arr, left, mid, right);

    return arr;
}


/**
 * Merge two sorted sections of an array in-place
 *
 * @param {number[]} arr - The array
 * @param {number} left - Start of left section
 * @param {number} mid - End of left section
 * @param {number} right - End of right section
 */
function mergeInPlace(arr, left, mid, right) {
    // Create temporary arrays
    const leftArr = arr.slice(left, mid + 1);
    const rightArr = arr.slice(mid + 1, right + 1);

    let i = 0;           // Index for leftArr
    let j = 0;           // Index for rightArr
    let k = left;        // Index for main array

    // Merge back into main array
    while (i < leftArr.length && j < rightArr.length) {
        if (leftArr[i] <= rightArr[j]) {
            arr[k] = leftArr[i];
            i++;
        } else {
            arr[k] = rightArr[j];
            j++;
        }
        k++;
    }

    // Copy remaining elements
    while (i < leftArr.length) {
        arr[k] = leftArr[i];
        i++;
        k++;
    }

    while (j < rightArr.length) {
        arr[k] = rightArr[j];
        j++;
        k++;
    }
}


/**
 * Bottom-up iterative merge sort (avoids recursion)
 *
 * @param {number[]} arr - Array to sort
 * @returns {number[]} - New sorted array
 */
function mergeSortIterative(arr) {
    if (arr.length <= 1) return arr;

    let result = [...arr];  // Work with a copy
    const n = result.length;

    // Start with size 1, double each iteration
    for (let size = 1; size < n; size *= 2) {
        // Merge subarrays of current size
        for (let start = 0; start < n; start += 2 * size) {
            const mid = Math.min(start + size - 1, n - 1);
            const end = Math.min(start + 2 * size - 1, n - 1);

            if (mid < end) {
                // Merge arr[start...mid] with arr[mid+1...end]
                const left = result.slice(start, mid + 1);
                const right = result.slice(mid + 1, end + 1);
                const merged = merge(left, right);

                // Copy back
                for (let i = 0; i < merged.length; i++) {
                    result[start + i] = merged[i];
                }
            }
        }
    }

    return result;
}


/**
 * Optimized merge sort with insertion sort for small arrays
 *
 * @param {number[]} arr - Array to sort
 * @param {number} threshold - Size threshold for insertion sort (default 10)
 * @returns {number[]} - New sorted array
 */
function mergeSortOptimized(arr, threshold = 10) {
    // Use insertion sort for small arrays
    if (arr.length <= threshold) {
        return insertionSort([...arr]);
    }

    const mid = Math.floor(arr.length / 2);
    const left = mergeSortOptimized(arr.slice(0, mid), threshold);
    const right = mergeSortOptimized(arr.slice(mid), threshold);

    // Optimization: if already sorted, just concatenate
    if (left[left.length - 1] <= right[0]) {
        return left.concat(right);
    }

    return merge(left, right);
}


/**
 * Simple insertion sort for small arrays
 *
 * @param {number[]} arr - Array to sort
 * @returns {number[]} - Sorted array
 */
function insertionSort(arr) {
    for (let i = 1; i < arr.length; i++) {
        const key = arr[i];
        let j = i - 1;

        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
    return arr;
}


/**
 * Merge sort with custom comparator
 *
 * @param {Array} arr - Array to sort
 * @param {Function} compareFn - Comparison function (a, b) => number
 * @returns {Array} - New sorted array
 */
function mergeSortComparator(arr, compareFn = (a, b) => a - b) {
    if (arr.length <= 1) return arr;

    const mid = Math.floor(arr.length / 2);
    const left = mergeSortComparator(arr.slice(0, mid), compareFn);
    const right = mergeSortComparator(arr.slice(mid), compareFn);

    return mergeWithComparator(left, right, compareFn);
}


/**
 * Merge with custom comparator
 *
 * @param {Array} left - First sorted array
 * @param {Array} right - Second sorted array
 * @param {Function} compareFn - Comparison function
 * @returns {Array} - Merged sorted array
 */
function mergeWithComparator(left, right, compareFn) {
    const result = [];
    let i = 0, j = 0;

    while (i < left.length && j < right.length) {
        if (compareFn(left[i], right[j]) <= 0) {
            result.push(left[i++]);
        } else {
            result.push(right[j++]);
        }
    }

    return result.concat(left.slice(i)).concat(right.slice(j));
}


/**
 * Merge sort that returns steps for visualization
 *
 * @param {number[]} arr - Array to sort
 * @returns {Object} - { result, steps }
 */
function mergeSortWithSteps(arr) {
    const steps = [];

    function mergeSortRecursive(array, offset = 0) {
        steps.push({
            type: 'divide',
            array: [...array],
            offset: offset,
            description: `Dividing array of length ${array.length}`
        });

        if (array.length <= 1) {
            return array;
        }

        const mid = Math.floor(array.length / 2);
        const left = mergeSortRecursive(array.slice(0, mid), offset);
        const right = mergeSortRecursive(array.slice(mid), offset + mid);

        const merged = mergeWithSteps(left, right, offset);

        steps.push({
            type: 'merge',
            array: merged,
            offset: offset,
            description: `Merged into array of length ${merged.length}`
        });

        return merged;
    }

    function mergeWithSteps(left, right, offset) {
        const result = [];
        let i = 0, j = 0;

        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) {
                result.push(left[i++]);
            } else {
                result.push(right[j++]);
            }
        }

        return result.concat(left.slice(i)).concat(right.slice(j));
    }

    const result = mergeSortRecursive(arr);

    return { result, steps };
}


// Example usage
if (require.main === module) {
    console.log('=== Merge Sort Examples ===\n');

    const arr1 = [38, 27, 43, 3, 9, 82, 10];
    console.log('Original:', arr1);
    console.log('Sorted:  ', mergeSort(arr1));

    console.log('\n--- In-Place Version ---');
    const arr2 = [64, 34, 25, 12, 22, 11, 90];
    console.log('Before:', arr2);
    mergeSortInPlace(arr2);
    console.log('After: ', arr2);

    console.log('\n--- Iterative Version ---');
    const arr3 = [5, 2, 8, 1, 9, 3, 7];
    console.log('Iterative:', mergeSortIterative(arr3));

    console.log('\n--- Optimized Version ---');
    const arr4 = [8, 3, 1, 7, 0, 10, 2, 4, 6, 5];
    console.log('Optimized:', mergeSortOptimized(arr4));

    console.log('\n--- Custom Comparator (Descending) ---');
    const arr5 = [5, 2, 8, 1, 9];
    console.log('Descending:', mergeSortComparator(arr5, (a, b) => b - a));

    console.log('\n--- Objects by Property ---');
    const people = [
        { name: 'Alice', age: 30 },
        { name: 'Bob', age: 25 },
        { name: 'Charlie', age: 35 }
    ];
    const sortedByAge = mergeSortComparator(people, (a, b) => a.age - b.age);
    console.log('By age:', sortedByAge.map(p => `${p.name}(${p.age})`).join(', '));
}


// Export for use in other files
module.exports = {
    mergeSort,
    merge,
    mergeSortInPlace,
    mergeSortIterative,
    mergeSortOptimized,
    mergeSortComparator,
    mergeSortWithSteps
};

File: algorithms/sorting/merge-sort/implementation.py
"""
Merge Sort Algorithm

Efficient divide-and-conquer sorting with guaranteed O(n log n) performance

Time: O(n log n) all cases
Space: O(n) auxiliary space
Stable: Yes
"""

from typing import List, TypeVar, Callable


T = TypeVar('T')


def merge_sort(arr: List[int]) -> List[int]:
    """
    Standard recursive merge sort (functional style).

    Args:
        arr: List of integers to sort

    Returns:
        New sorted list

    Examples:
        >>> merge_sort([38, 27, 43, 3, 9, 82, 10])
        [3, 9, 10, 27, 38, 43, 82]

        >>> merge_sort([5, 2, 8, 1, 9])
        [1, 2, 5, 8, 9]

        >>> merge_sort([1])
        [1]

        >>> merge_sort([])
        []

        >>> merge_sort([5, 5, 5])
        [5, 5, 5]
    """
    # Base case: arrays of 0 or 1 element are already sorted
    if len(arr) <= 1:
        return arr

    # Divide: split array in half
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]

    # Conquer: recursively sort both halves
    sorted_left = merge_sort(left)
    sorted_right = merge_sort(right)

    # Combine: merge the sorted halves
    return merge(sorted_left, sorted_right)


def merge(left: List[int], right: List[int]) -> List[int]:
    """
    Merge two sorted arrays into one sorted array.

    Args:
        left: First sorted list
        right: Second sorted list

    Returns:
        Merged sorted list

    Examples:
        >>> merge([1, 3, 5], [2, 4, 6])
        [1, 2, 3, 4, 5, 6]

        >>> merge([1], [2])
        [1, 2]

        >>> merge([], [1, 2, 3])
        [1, 2, 3]
    """
    result = []
    i = j = 0

    # Compare elements from left and right, add smaller to result
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:  # <= ensures stability
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # Copy any remaining elements
    result.extend(left[i:])
    result.extend(right[j:])

    return result


def merge_sort_in_place(arr: List[int], left: int = 0, right: int = None) -> List[int]:
    """
    In-place merge sort (modifies original array).

    Args:
        arr: List to sort
        left: Left boundary (default 0)
        right: Right boundary (default len(arr) - 1)

    Returns:
        The same list, sorted

    Examples:
        >>> arr = [38, 27, 43, 3, 9, 82, 10]
        >>> merge_sort_in_place(arr)
        [3, 9, 10, 27, 38, 43, 82]

        >>> arr = [5, 4, 3, 2, 1]
        >>> merge_sort_in_place(arr)
        [1, 2, 3, 4, 5]
    """
    if right is None:
        right = len(arr) - 1

    if left >= right:
        return arr

    mid = (left + right) // 2

    # Sort left and right halves
    merge_sort_in_place(arr, left, mid)
    merge_sort_in_place(arr, mid + 1, right)

    # Merge the sorted halves
    _merge_in_place(arr, left, mid, right)

    return arr


def _merge_in_place(arr: List[int], left: int, mid: int, right: int) -> None:
    """
    Merge two sorted sections of an array in-place.

    Args:
        arr: The list
        left: Start of left section
        mid: End of left section
        right: End of right section
    """
    # Create temporary arrays
    left_arr = arr[left:mid + 1]
    right_arr = arr[mid + 1:right + 1]

    i = j = 0
    k = left

    # Merge back into main array
    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1

    # Copy remaining elements
    while i < len(left_arr):
        arr[k] = left_arr[i]
        i += 1
        k += 1

    while j < len(right_arr):
        arr[k] = right_arr[j]
        j += 1
        k += 1


def merge_sort_iterative(arr: List[int]) -> List[int]:
    """
    Bottom-up iterative merge sort (avoids recursion).

    Args:
        arr: List to sort

    Returns:
        New sorted list

    Examples:
        >>> merge_sort_iterative([5, 2, 8, 1, 9, 3, 7])
        [1, 2, 3, 5, 7, 8, 9]

        >>> merge_sort_iterative([1, 2, 3])
        [1, 2, 3]
    """
    if len(arr) <= 1:
        return arr

    result = arr.copy()
    n = len(result)

    # Start with size 1, double each iteration
    size = 1
    while size < n:
        # Merge subarrays of current size
        start = 0
        while start < n:
            mid = min(start + size - 1, n - 1)
            end = min(start + 2 * size - 1, n - 1)

            if mid < end:
                # Merge result[start...mid] with result[mid+1...end]
                left = result[start:mid + 1]
                right = result[mid + 1:end + 1]
                merged = merge(left, right)

                # Copy back
                for i, val in enumerate(merged):
                    result[start + i] = val

            start += 2 * size
        size *= 2

    return result


def merge_sort_generic(arr: List[T], key: Callable[[T], any] = None) -> List[T]:
    """
    Generic merge sort with optional key function.

    Args:
        arr: List to sort
        key: Optional key function for comparison

    Returns:
        New sorted list

    Examples:
        >>> merge_sort_generic([5, 2, 8, 1, 9])
        [1, 2, 5, 8, 9]

        >>> merge_sort_generic([5, 2, 8, 1, 9], key=lambda x: -x)
        [9, 8, 5, 2, 1]

        >>> people = [('Alice', 30), ('Bob', 25), ('Charlie', 35)]
        >>> merge_sort_generic(people, key=lambda x: x[1])
        [('Bob', 25), ('Alice', 30), ('Charlie', 35)]
    """
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort_generic(arr[:mid], key)
    right = merge_sort_generic(arr[mid:], key)

    return _merge_generic(left, right, key)


def _merge_generic(left: List[T], right: List[T], key: Callable[[T], any] = None) -> List[T]:
    """
    Merge two sorted arrays with optional key function.

    Args:
        left: First sorted list
        right: Second sorted list
        key: Optional key function for comparison

    Returns:
        Merged sorted list
    """
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        left_val = key(left[i]) if key else left[i]
        right_val = key(right[j]) if key else right[j]

        if left_val <= right_val:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])

    return result


if __name__ == '__main__':
    print('=== Merge Sort Examples ===\n')

    # Basic example
    arr1 = [38, 27, 43, 3, 9, 82, 10]
    print(f'Original: {arr1}')
    print(f'Sorted:   {merge_sort(arr1)}')

    # In-place version
    print('\n--- In-Place Version ---')
    arr2 = [64, 34, 25, 12, 22, 11, 90]
    print(f'Before: {arr2}')
    merge_sort_in_place(arr2)
    print(f'After:  {arr2}')

    # Iterative version
    print('\n--- Iterative Version ---')
    arr3 = [5, 2, 8, 1, 9, 3, 7]
    print(f'Iterative: {merge_sort_iterative(arr3)}')

    # Generic with key
    print('\n--- Descending Order ---')
    arr4 = [5, 2, 8, 1, 9]
    print(f'Descending: {merge_sort_generic(arr4, key=lambda x: -x)}')

    # Objects by property
    print('\n--- Objects by Property ---')
    people = [
        ('Alice', 30),
        ('Bob', 25),
        ('Charlie', 35)
    ]
    sorted_by_age = merge_sort_generic(people, key=lambda x: x[1])
    print(f'By age: {sorted_by_age}')

    # Run doctests
    print('\n--- Running Doctests ---')
    import doctest
    results = doctest.testmod()
    if results.failed == 0:
        print(f'✓ All {results.attempted} doctests passed!')
    else:
        print(f'✗ {results.failed} doctest(s) failed')
